bitsail-connectors/connector-assert/src/main/java/com/bytedance/bitsail/connector/assertion/sink/AssertWriter.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2022 Bytedance Ltd. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.bytedance.bitsail.connector.assertion.sink;
+
+import com.bytedance.bitsail.base.connector.writer.v1.Writer;
+import com.bytedance.bitsail.common.BitSailException;
+import com.bytedance.bitsail.common.row.Row;
+import com.bytedance.bitsail.connector.assertion.sink.error.AssertErrorCode;
+import com.bytedance.bitsail.connector.assertion.sink.executor.AssertRuleExecutor;
+import com.bytedance.bitsail.connector.assertion.sink.rule.AssertColumnRule;
+import com.bytedance.bitsail.connector.assertion.sink.rule.AssertSimpleRule;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class AssertWriter implements Writer<Row, String, Integer> {
+
+  private final List<AssertSimpleRule> rowRules;
+
+  private final Map<String, AssertColumnRule> columnRules;
+
+  private final List<String> columnNames;
+
+  private final AtomicLong atomicLong = new AtomicLong(0);
+
+  private final AssertRuleExecutor assertRuleExecutor = new AssertRuleExecutor();
+
+  public AssertWriter(List<AssertSimpleRule> rowRules, Map<String, AssertColumnRule> columnRules, List<String> columnNames) {
+    this.rowRules = rowRules;
+    this.columnRules = columnRules;
+    this.columnNames = columnNames;
+  }
+
+  @Override
+  public void write(Row element) throws IOException {
+    atomicLong.getAndIncrement();
+    if (columnRules != null) {
+      AssertColumnRule rule = assertRuleExecutor.check(element, columnRules, columnNames);
+      if (rule != null) {
+        throw BitSailException.asBitSailException(
+            AssertErrorCode.RULE_CHECK_FAILED, "Row : " + element + " check failed rule: " + rule);
+      }
+    }
+  }
+
+  @Override
+  public void flush(boolean endOfInput) throws IOException {
+
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (rowRules != null) {
```

Comment:
didn't get this. Why do we do this at closing

Reply:
`rowRules` is used to limit the number of processed data by engine, we need to ensure that the number of processed data is within the range of `[min_row, max_row]`. In `write` method, we add counter the after handler a row and compare it in `close` method finally.

Reply:
so the row rule must be a single parallelism right, looks like this counter only count the number of rows in this writer.

Reply:
thanks for pointing this out, i think so, if we want to limit row number, the parallelism must be one, can we set it `writer_parallelism_num` by this conf, or any suggestions, thx

Reply:
if the assert sink only support single parallelism, we should make sure the single parallelism was hard coded to ensure the correct result. The related class is `ParallelismComputable`. 
If we wanna support multi parallelism, I believe we should some how aggregate the count in the coordinator. We could make the multi parallelism support as a follow up task for future work.

Reply:
Thanks so much, got it, i'll look `ParallelismComputable` first.